### PR TITLE
Ji2578 - Tiers Variable Context

### PR DESCRIPTION
**Pull request by jericmason**

### CODE REVIEW

File: src/json/ProgramRequest.schema.json [hunk 1]
```diff
@@ -21,7 +21,8 @@
       "oneOf": [
         { "$ref": "#/definitions/programTrigger" },
         { "$ref": "#/definitions/programValidation" },
-        { "$ref": "#/definitions/programIntrospection" }
+        { "$ref": "#/definitions/programIntrospection" },
+        { "$ref": "#/definitions/programTriggerVariablesSchemaRequest" }
```
Comment (loganvolkers) [inline src/json/ProgramRequest.schema.json]: @jericmason I'd like to read the behaviour specs for this, understand what it's used for

File: src/json/program/AfterUserCreatedOrUpdatedTriggerContext.schema.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  type: "object",
+  title: "After User Created Or Updated Trigger Context",
+  description:
+    "Defines the trigger variables available to an AFTER_USER_CREATED_OR_UPDATED trigger",
+  properties: {
+    type: {
+      enum: ["AFTER_USER_CREATED_OR_UPDATED"]
+    },
+    user: {
+      title: "Current User Data",
```
Comment (loganvolkers) [inline src/json/program/AfterUserCreatedOrUpdatedTriggerContext.schema.ts]:
```suggestion
      title: "Current User",
```

File: src/json/program/AfterUserCreatedOrUpdatedTriggerContext.schema.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      ...userContext
+    },
+    previous: {
+      title: "Previous User Data",
```
Comment (loganvolkers) [inline src/json/program/AfterUserCreatedOrUpdatedTriggerContext.schema.ts]:
```suggestion
      title: "Previous User",
```

Comment (loganvolkers) [reply src/json/program/AfterUserCreatedOrUpdatedTriggerContext.schema.ts]: @jorgecgll and @rebeccachan looking for your feedback on the labels for these tier variables.

These labels will show up in some way in the portal (for picking custom fields for example), in our docs (similar to our existing Theme Variables used in handlebars), in logs, error messages, etc.

Important to get them right

File: src/json/program/AfterUserEventProcessedTriggerContext.schema.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  type: "object",
+  title: "After User Event Processed Trigger Context",
+  description:
+    "Defines the trigger variables available to an AFTER_USER_EVENT_PROCESSED trigger",
+  properties: {
+    type: {
+      enum: ["AFTER_USER_EVENT_PROCESSED"]
+    },
+    user: {
+      title: "Current User Data",
```
Comment (loganvolkers) [inline src/json/program/AfterUserEventProcessedTriggerContext.schema.ts]: I think `current` is implied

```suggestion
      title: "User Data",
```

File: src/json/program/AfterUserEventProcessedTriggerContext.schema.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import {
+  contextTimeField,
+  userContext,
+  customFieldsAndEventsDefinitions
+} from "./ProgramTriggerVariableContext.schema";
+
+const afterUserEventProcessedTriggerContextSchema: JSONSchema6 = {
+  $schema: "http://json-schema.org/draft-06/schema#",
+  type: "object",
+  title: "After User Event Processed Trigger Context",
```
Comment (loganvolkers) [inline src/json/program/AfterUserEventProcessedTriggerContext.schema.ts]: Make it more doc friendly?

```suggestion
  title: "Event Trigger Variables",
```

File: src/json/program/AfterUserEventProcessedTriggerContext.schema.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  description:
+    "Defines the trigger variables available to an AFTER_USER_EVENT_PROCESSED trigger",
```
Comment (loganvolkers) [inline src/json/program/AfterUserEventProcessedTriggerContext.schema.ts]: Make this more humanized / doc friendly

```suggestion
    "These variables are available for designing conversion rules for Event triggers",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+    }
+  }
+};
+
+const programTriggerVariableContextSchema2: JSONSchema6 = {
+  $schema: "http://json-schema.org/draft-06/schema#",
+  type: "object",
+  title: "Program Variable Context Schema",
+  description:
+    "Defines the context used for evaluating externally defined inputs (e.g. JSONata)",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
    "The variables used for evaluating externally defined inputs (e.g. JSONata)",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 8]
```diff
@@ -0,0 +1,396 @@
+import { JSONSchema6 } from "json-schema";
+
+const programTriggerVariableContextSchema: JSONSchema6 = {
+  $schema: "http://json-schema.org/draft-06/schema#",
+  type: "object",
+  title: "Program Variable Context Schema",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
  title: "Program Trigger Variables",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+  description:
+    "Defines the context used for evaluating externally defined inputs (e.g. JSONata)",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
    "Defines the variables used for evaluating externally defined inputs (e.g. JSONata)",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+        "AFTER_USER_EVENT_PROCESSED"
+      ]
+    }
+  }
+};
+
+const programTriggerVariableContextSchema2: JSONSchema6 = {
+  $schema: "http://json-schema.org/draft-06/schema#",
+  type: "object",
+  title: "Program Variable Context Schema",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
  title: "Program Trigger Variables",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+    type: ["string", "null"],
+    title: "Email Address"
+  },
+  cookieId: {
+    type: ["string", "null"],
+    title: "Cookie Id"
+  },
+  locale: {
+    type: ["string", "null"],
+    title: "User Locale"
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]: The `user` part is implied I think

```suggestion
    title: "Locale"
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+  },
+  referable: {
+    type: "boolean",
+    title: "Is Referable"
+  },
+  firstSeenIP: {
+    type: ["string", "null"],
+    title: "First Seen Ip"
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]: @jorgecgll Is this consistent with the portal?

```suggestion
    title: "First Seen IP Address"
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+  },
+  lastSeenIP: {
+    type: ["string", "null"],
+    title: "Last Seen Ip"
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
    title: "Last Seen IP Address"
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+  },
+  firstSeenGeoData: {
+    type: "object",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]: @jericmason The sub-values for this either need to be included, or this should be removed from the context.

It's rather useless for docs/frontend/validation/etc. without some details about what's inside.

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+  dateBlocked: {
+    type: ["integer", "null"],
+    title: "Date User Blocked"
+  },
+  customFields: {
+    $ref: "#/definitions/customFieldsContext"
+  },
+  segments: {
+    type: "array",
+    title: "User Segements",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
    title: "Segments",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+    description: "A list of segments this user has been assigned to",
+    items: {
+      type: "string",
+      title: "User Segment",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
      title: "Segment",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+      description: "A segment this user has been assigned to (e.g. VIP)"
+    }
+  },
+  fraudFlags: {
+    type: "array",
+    title: "User Fraud Flags",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
    title: "Fraud Flags",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+    description:
+      "Flags indicating whether this user was marked as a potentially fradulent user",
+    items: {
+      type: "object",
+      title: "Fraud Flags",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
      title: "Fraud Flag",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+    },
+    dateCancelled: {
+      type: ["string", "null"],
+      title: "Date Cancelled",
+      description: "The date the reward was cancelled"
+    },
+    programId: {
+      type: ["string", "null"],
+      title: "Program Id",
+      description: "The identifier or the program that generated this reward"
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
      description: "The identifier of the program that generated this reward"
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+  }
+};
+
+const customFieldsAndEventsDefinitions: JSONSchema6 = {
+  definitions: {
+    eventContext: {
+      type: "object",
+      title: "Event",
+      description:
+        "The context of the event that triggered the program and is currently being evaluated",
```
Comment (loganvolkers) [inline src/json/program/ProgramTriggerVariableContext.schema.ts]:
```suggestion
        "The variables for the event that triggered the program",
```

File: src/json/program/RewardScheduledProgramTriggerContext.schema.ts [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+import {
+  contextTimeField,
+  userContext,
+  customFieldsAndEventsDefinitions
+} from "./ProgramTriggerVariableContext.schema";
+
+const rewardScheduledProgramTriggerContextSchema: JSONSchema6 = {
+  $schema: "http://json-schema.org/draft-06/schema#",
+  type: "object",
+  title: "Reward Scheduled Context",
```
Comment (loganvolkers) [inline src/json/program/RewardScheduledProgramTriggerContext.schema.ts]:
```suggestion
  title: "Reward Scheduled Trigger Variables",
```

File: src/json/program/ScheduledProgramTriggerContext.schema.ts [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+import {
+  contextTimeField,
+  userContext,
+  customFieldsAndEventsDefinitions
+} from "./ProgramTriggerVariableContext.schema";
+
+const scheduledProgramTriggerContextSchema: JSONSchema6 = {
+  $schema: "http://json-schema.org/draft-06/schema#",
+  type: "object",
+  title: "Scheduled Context",
```
Comment (loganvolkers) [inline src/json/program/ScheduledProgramTriggerContext.schema.ts]:
```suggestion
  title: "Scheduled Trigger Variables",
```

**Comment by loganvolkers** [comment]
@jericmason and @jorgecgll my biggest concerns here are:

 - **Consistency** with the UI and other docs. e.g. do we call it `Image URL` in one place and `Profile Picture` in another?
 - **Clarity** of descriptions. 
   - When we should this in the UI, how do we represent `User ➝ Custom Fields ➝ LeadStage` vs `Previous User ➝ Custom Fields ➝ LeadStage`. I think the current UI might just show `LeadStage`, but that's ambiguous. 
   - If we show the full path, then the `title` fields for arrays and objects become important. e.g. `User ➝ Locale` vs `User ➝ User Locale`.
   - We might want to copy Salesforce and use the arrow style, e.g. `User ➝ First Name` 
   - Links for arrow emoticons: https://fsymbols.com/signs/arrow/
 - **Examples**. It might be useful for there to be `example` values in the schema, docs and UI. `locale` and `rewardKey` come to mind for having good example values.

File: src/json/program/RewardScheduledProgramTriggerContext.schema.ts [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+  description: "Defines the trigger variables available to a REWARD_SCHEDULED trigger",
+  properties: {
+    type: {
+      enum: ["REWARD_SCHEDULED"]
+    },
+    user: {
+      title: "Current User Data",
```
Comment (loganvolkers) [inline src/json/program/RewardScheduledProgramTriggerContext.schema.ts]:
```suggestion
      title: "User",
```

File: src/json/program/ProgramTriggerVariableContext.schema.ts (hunk 14 above)
Comment (jericmason) [reply src/json/program/ProgramTriggerVariableContext.schema.ts]: Removed the geo data fields for now, will re-add later when we have the full schema for them.

**Comment by loganvolkers** [comment]
@jericmason any reason why this hasn't been merged?

**Comment by jericmason** [comment]
Merging this as it has already been deployed.

**Pull request merged**